### PR TITLE
add unit tests for models/challenges and add time utils 

### PR DESCRIPTION
**Pull request by pallabez**
Helps Fixes 
- https://github.com/Real-Dev-Squad/website-backend/issues/539

## Todo 
- Fix data model of challenges

## New test coverage
image

### CODE REVIEW

File: test/fixtures/challenges/challenges.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+module.exports = () => {
+  return [
+    {
+      title: "Sherlock and Anagrams",
+      level: "Easy",
+      start_date: parseInt(Date.now() / 1000),
+      end_date: parseInt(Date.now() / 1000) + 60 * 60 * 24 * 10, // After 10 days
```
Comment (shubham-y) [inline test/fixtures/challenges/challenges.js:7]: **Suggestion**: Add a constant variable to store seconds
Reference - https://github.com/ryanmcdermott/clean-code-javascript#use-searchable-names

File: test/unit/models/challenges.test.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
+/* eslint-disable security/detect-object-injection */
```
Comment (shubham-y) [inline test/unit/models/challenges.test.js:1]: Please check if we can remove this

File: test/unit/models/challenges.test.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+const challengeDataArray = require("../../fixtures/challenges/challenges")();
+const challengeData = challengeDataArray[0];
+
+describe("Challenges", function () {
+  afterEach(async function () {
+    await cleanDb();
+  });
+  describe("postChallenge", function () {
+    it("should add the challenge", async function () {
+      const challengeId = await challengeQuery.postChallenge(challengeData);
```
Comment (shubham-y) [inline test/unit/models/challenges.test.js:23]: Can we move it to `beforeEach` function as it's being used in multiple it blocks?

**Review by shubham-y — CHANGES_REQUESTED** [review]
Good work
Requesting some changes

File: test/fixtures/challenges/challenges.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+module.exports = () => {
+  return [
+    {
+      title: "Sherlock and Anagrams",
+      level: "Easy",
+      start_date: parseInt(Date.now() / 1000),
+      end_date: parseInt(Date.now() / 1000) + convertToSeconds.DAYS_10, // After 10 days
+    },
+  ];
```
Comment (ankushdharkar) [inline test/fixtures/challenges/challenges.js:11]: Instead of doing mathematical calculations here, I am wondering if we can import a util function for our Date needs?

```js
getTimeNow()
getTimeAfter({ days: 10})
```

Also, why are we passing timestamp in a "<>_date" key?

File: test/unit/models/challenges.test.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    });
+    it("should return a empty array", async function () {
+      await cleanDb();
+      const response = await challengeQuery.fetchChallenges();
+
+      expect(response).to.be.a("array").with.lengthOf(0);
+    });
+  });
+
+  describe("subscribeUserToChallenge", function () {
```
Comment (shubham-y) [inline test/unit/models/challenges.test.js:60]: We should add a test passing invalid userId to `subscribeUserToChallenge()` and `fetchParticipantsData()`. What do you think?

Comment (shubham-y) [reply test/unit/models/challenges.test.js:60]: Also, pass a challenge fixture that already has a participant in order to test multiple cases

Comment (pallabez) [reply test/unit/models/challenges.test.js:61]: Done! :+1:

File: test/fixtures/challenges/challenges.js (hunk 4 above)
Comment (pallabez) [reply test/fixtures/challenges/challenges.js]: Done :+1:

File: test/fixtures/time/time.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
+const minutesToMilliseconds = [
```
Comment (shubham-y) [inline test/fixtures/time/time.js:1]: Suggestion to please use capital letters and underscores for naming constant variables
Reference - https://javascript.info/variables#uppercase-constants

File: test/unit/utils/time.test.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
+/* eslint-disable security/detect-object-injection */
```
Comment (shubham-y) [inline test/unit/utils/time.test.js:1]: Please check if we can remove this

File: models/challenges.js [hunk 8]
```diff
@@ -66,13 +67,13 @@ const fetchParticipantsData = async (participants) => {
...
+    const endat = new Firestore.Timestamp(getTimeInSecondAfter({ timestamp: endAt }), 0);
     const challengeRef = await challengesModel.add({
       ...challengeData,
-      start_date: startdate,
-      end_date: enddate,
+      start_at: startat,
+      end_at: endat,
       participants: [],
       is_active: true,
     });
```
Comment (shubham-y) [inline models/challenges.js:79]: I think it will be better to raise a different PR for changing the keys of challenges object because it might cause issue in frontend. What do you think?

Comment (pallabez) [reply models/challenges.js:79]: Yeah, you are right. Will do that.

File: test/fixtures/time/time.js (hunk 6 above)
Comment (pallabez) [reply test/fixtures/time/time.js:1]: No function in fixtures follows it. I believe it's not really a constant. Is it still suggested?

File: test/unit/utils/time.test.js (hunk 7 above)
Comment (pallabez) [reply test/unit/utils/time.test.js:1]: Yes. Thanks for pointing that out.
